File: src/modules/npm.ts
import { existsSync } from 'fs';
import { join } from 'path';
import type { AnalysisResult, CleaningResult } from '../types/index.js';
import { getHomeDir } from '../utils/os.js';
import { exec, isCommandAvailable } from '../utils/exec.js';
import { logger } from '../utils/logger.js';
import { getDirSizeAsync } from '../utils/fs.js';

abstract class PackageCacheModule {
  abstract readonly id: string;
  abstract readonly name: string;
  abstract readonly description: string;
  abstract getCachePaths(): string[];

  isAvailable(): boolean {
    return isCommandAvailable(this.id);
  }

  async analyze(): Promise<AnalysisResult> {
    const items: AnalysisResult['items'] = [];

    const results = await Promise.all(
      this.getCachePaths().map(async cachePath => {
        if (!existsSync(cachePath)) return null;
        try {
          const size = await getDirSizeAsync(cachePath);
          return { cachePath, size };
        } catch {
          return null;
        }
      })
    );

    let totalSize = 0;
    for (const result of results) {
      if (result) {
        items.push({
          path: result.cachePath,
          size: result.size,
          type: 'directory',
          description: `Cache do ${this.name} (${result.cachePath})`,
        });
        totalSize += result.size;
      }
    }

    return { module: this.id, items, totalSize };
  }

  abstract getCleanCommand(): string[];

  async clean(dryRun: boolean = false): Promise<CleaningResult> {
    const analysis = await this.analyze();
    const result: CleaningResult = {
      module: this.id,
      success: true,
      spaceFreed: 0,
      itemsRemoved: 0,
      errors: [],
    };

    if (analysis.totalSize === 0) {
      return result;
    }

    if (dryRun) {
      logger.info(
        `[DRY-RUN] Limparía ${logger.formatBytes(analysis.totalSize)} do cache ${this.name.toLowerCase()}`
      );
      result.spaceFreed = analysis.totalSize;
      return result;
    }

    const commandResult = await exec(this.id, this.getCleanCommand());

    if (commandResult.success) {
      result.success = true;
      result.spaceFreed = analysis.totalSize;
      result.itemsRemoved = 1;
      logger.item(`${this.name}: Cache limpo`, logger.formatBytes(analysis.totalSize));
    } else {
      result.success = false;
      result.errors.push(commandResult.stderr || `Falha ao limpar cache ${this.name}`);
    }

    return result;
  }
}

export class NpmModule extends PackageCacheModule {
  readonly id = 'npm';
  readonly name = 'NPM';
  readonly description = 'Limpa cache do npm';

  getCachePaths(): string[] {
    return [join(getHomeDir(), '.npm')];
  }

  getCleanCommand(): string[] {
    return ['cache', 'clean', '--force'];
  }
}

export class YarnModule extends PackageCacheModule {
  readonly id = 'yarn';
  readonly name = 'Yarn';
  readonly description = 'Limpa cache do Yarn';

  getCachePaths(): string[] {
    return [join(getHomeDir(), '.cache', 'yarn'), join(getHomeDir(), '.yarn', 'cache')];
  }

  getCleanCommand(): string[] {
    return ['cache', 'clean'];
  }
}

export class PnpmModule extends PackageCacheModule {
  readonly id = 'pnpm';
  readonly name = 'PNPM';
  readonly description = 'Limpa cache do PNPM';

  getCachePaths(): string[] {
    const paths = [join(getHomeDir(), '.pnpm-store')];
    const localPath = join(getHomeDir(), '.local', 'share', 'pnpm', 'cache');
    if (existsSync(localPath)) {
      paths.push(localPath);
    }
    return paths;
  }

  getCleanCommand(): string[] {
    return ['store', 'prune'];
  }
}

export const npmModule = new NpmModule();
export const yarnModule = new YarnModule();
export const pnpmModule = new PnpmModule();

File: internal/modules/npm.go
package modules

import (
	"fmt"
	"path/filepath"
	"strings"

	"github.com/joaomjbraga/piunter/internal/utils"
	"github.com/joaomjbraga/piunter/pkg/types"
)

type PackageCacheModule struct {
	BaseModule
	cachePaths    []func() []string
	cleanCommands []string
}

func (m *PackageCacheModule) IsAvailable() bool {
	for _, getPaths := range m.cachePaths {
		for _, path := range getPaths() {
			if utils.FileExists(path) {
				return true
			}
		}
	}
	return false
}

func (m *PackageCacheModule) Analyze(threshold int) (*types.AnalysisResult, error) {
	result := &types.AnalysisResult{
		Module:    m.ID(),
		Items:     []types.CleanableItem{},
		TotalSize: 0,
	}

	for _, getPaths := range m.cachePaths {
		for _, path := range getPaths() {
			if !utils.FileExists(path) {
				continue
			}
			size := utils.GetDirSizeAsync(path)
			result.Items = append(result.Items, types.CleanableItem{
				Path:        path,
				Size:        size,
				Type:        "directory",
				Description: fmt.Sprintf("Cache do %s (%s)", m.Name(), path),
			})
			result.TotalSize += size
		}
	}

	return result, nil
}

func (m *PackageCacheModule) Clean(dryRun bool) (*types.CleaningResult, error) {
	analysis, err := m.Analyze(0)
	if err != nil {
		return &types.CleaningResult{
			Module:  m.ID(),
			Success: false,
			Errors:  []string{err.Error()},
		}, err
	}

	result := &types.CleaningResult{
		Module:       m.ID(),
		Success:      true,
		SpaceFreed:   0,
		ItemsRemoved: 0,
		Errors:       []string{},
	}

	if analysis.TotalSize == 0 {
		return result, nil
	}

	if dryRun {
		result.SpaceFreed = analysis.TotalSize
		utils.Info(fmt.Sprintf("[DRY-RUN] Limparia %s do cache %s", utils.FormatBytes(analysis.TotalSize), m.Name()))
		return result, nil
	}

	for _, cmd := range m.cleanCommands {
		parts := strings.Fields(cmd)
		execResult := utils.Exec(parts[0], parts[1:]...)
		if execResult.Success {
			result.SpaceFreed = analysis.TotalSize
			result.ItemsRemoved = 1
			utils.Item(m.Name(), "Cache limpo")
		} else {
			result.Success = false
			result.Errors = append(result.Errors, fmt.Sprintf("Falha ao limpar cache %s: %s", m.Name(), execResult.Stderr))
		}
	}

	return result, nil
}

type NpmModule struct {
	PackageCacheModule
}

func NewNpmModule() *NpmModule {
	home := utils.GetHomeDir()
	return &NpmModule{
		PackageCacheModule: PackageCacheModule{
			BaseModule: BaseModule{
				id:          "npm",
				name:        "NPM",
				description: "Limpa cache do npm",
			},
			cachePaths: []func() []string{
				func() []string { return []string{filepath.Join(home, ".npm")} },
			},
			cleanCommands: []string{"npm cache clean --force"},
		},
	}
}

type YarnModule struct {
	PackageCacheModule
}

func NewYarnModule() *YarnModule {
	home := utils.GetHomeDir()
	return &YarnModule{
		PackageCacheModule: PackageCacheModule{
			BaseModule: BaseModule{
				id:          "yarn",
				name:        "Yarn",
				description: "Limpa cache do Yarn",
			},
			cachePaths: []func() []string{
				func() []string { return []string{filepath.Join(home, ".cache", "yarn"), filepath.Join(home, ".yarn", "cache")} },
			},
			cleanCommands: []string{"yarn cache clean"},
		},
	}
}

type PnpmModule struct {
	PackageCacheModule
}

func NewPnpmModule() *PnpmModule {
	home := utils.GetHomeDir()
	return &PnpmModule{
		PackageCacheModule: PackageCacheModule{
			BaseModule: BaseModule{
				id:          "pnpm",
				name:        "PNPM",
				description: "Limpa cache do PNPM",
			},
			cachePaths: []func() []string{
				func() []string {
					paths := []string{filepath.Join(home, ".pnpm-store")}
					localPath := filepath.Join(home, ".local", "share", "pnpm", "cache")
					if utils.FileExists(localPath) {
						paths = append(paths, localPath)
					}
					return paths
				},
			},
			cleanCommands: []string{"pnpm store prune"},
		},
	}
}